batcher/sweep_batch.go

```diff
@@ -0,0 +1,1330 @@
+			),
+		)
+		if err != nil {
+			return err
+		}
+
+		if b.muSig2SignSweep == nil {
+			return fmt.Errorf("no musig2 signer available")
+		}
+
+		// With the session active, we can now send the server our public nonce
+		// and the sig hash, so that it can create it's own MuSig2 session and
+		// return the server side nonce and partial signature.
+		serverNonce, serverSig, err := b.muSig2SignSweep(
+			ctx, sweep.ProtocolVersion, sweep.SwapHash,
+			sweep.SwapInvoicePaymentAddr,
+			musig2SessionInfo.PublicNonce[:], psbtBuf.Bytes(),
+			prevOuts,
+		)
+		if err != nil {
+			return err
+		}
+
+		var serverPublicNonce [musig2.PubNonceSize]byte
+		copy(serverPublicNonce[:], serverNonce)
+
+		// Register the server's nonce before attempting to create our partial
+		// signature.
+		haveAllNonces, err := b.signerClient.MuSig2RegisterNonces(
+			ctx, musig2SessionInfo.SessionID,
+			[][musig2.PubNonceSize]byte{serverPublicNonce},
+		)
+		if err != nil {
+			return err
+		}
+
+		// Sanity check that we have all the nonces.
+		if !haveAllNonces {
+			return fmt.Errorf("invalid MuSig2 session: " +
+				"nonces missing")
+		}
+
+		var digest [32]byte
+		copy(digest[:], sigHash)
+
+		// Since our MuSig2 session has all nonces, we can now create the local
+		// partial signature by signing the sig hash.
+		_, err = b.signerClient.MuSig2Sign(
+			ctx, musig2SessionInfo.SessionID, digest, false,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Now combine the partial signatures to use the final combined
+		// signature in the sweep transaction's witness.
+		haveAllSigs, finalSig, err := b.signerClient.MuSig2CombineSig(
+			ctx, musig2SessionInfo.SessionID, [][]byte{serverSig},
+		)
+		if err != nil {
+			return err
+		}
+
+		if !haveAllSigs {
+			return fmt.Errorf("failed to combine signatures")
+		}
+
+		// To be sure that we're good, parse and validate that the combined
+		// signature is indeed valid for the sig hash and the internal pubkey.
+		err = b.verifySchnorrSig(
+			htlcScript.TaprootKey, sigHash, finalSig,
+		)
+		if err != nil {
+			return err
+		}
+
+		packet.UnsignedTx.TxIn[i].Witness = wire.TxWitness{
+			finalSig,
+		}
+	}
+
+	return nil
+}
+
+// updateRbfRate updates the fee rate we should use for the new batch
+// transaction. This fee rate does not guarantee RBF success, but the continuous
+// increase leads to an eventual successful RBF replacement.
+func (b *Batch) updateRbfRate(ctx context.Context) error {
+	// If the feeRate is unset then we never published before, so we
+	// retrieve the fee estimate from our wallet.
+	if b.RbfCache.FeeRate == 0 {
+		b.log.Infof("initializing rbf fee rate for conf target=%v", b.Cfg.BatchConfTarget)
+		rate, err := b.wallet.EstimateFeeRate(
+			ctx, b.Cfg.BatchConfTarget,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Set the initial value for our fee rate.
+		b.RbfCache.FeeRate = rate
+	} else {
+		// Bump the fee rate by the configured step.
+		b.RbfCache.FeeRate += defaultFeeRateStep
```

Comment:
A better way to do this would be to also ask an estimate and see which one is higher, the stepped or the estimate. Otherwise the RBF cache won't react to fee rate changes in the mempool.

Reply:
We want to keep a low-balling approach in order to not spend too much because of temporary spikes